### PR TITLE
Use envrironment vars for recovery and oc password

### PR DESCRIPTION
**Pull request by sharidas**
Instead of providing password for all the users, be
it recovery or oc password, its better to read it
from the environment variable. This change helps
users to read it from the environment variable.
If user wishes to provide oc password for each user,
that option is also available.

Signed-off-by: Sujith H <sharidasan@owncloud.com>

## Related Issue

- Fixes <issue_link>

## How Has This Been Tested?

- create 2 users `admin` and `user1`
- Enable encryption with user-keys. Login as `admin` and `user1`.
- Now try to run decrypt-all command as follows:
1. `./occ encryption:decrypt-all -c yes` - This will decrypt both `admin` and `user1`. Confirmation questions will be asked, like whether to continue or like enter login password. This step shows us that old behaviour of entering oc password is preserved. Here is the console output log:
```
➜  owncloud3 git:(recovery_passwd-and-oc_passwd-from-env) ✗ ./occ encryption:decrypt-all -c yes            
Cannot load Xdebug - it was already loaded
Disable server side encryption... done.


You are about to start to decrypt all files stored in your ownCloud.
It will depend on the encryption module and your setup if this is possible.
Depending on the number and size of your files this can take some time
Please make sure that no user access his files during this process!

prepare encryption modules...
 done.


 starting to decrypt files... 
 [->--------------------------]
Prepare "Default encryption module"

Configuring encryption module for decryption with user based keys
Please enter the user: admin login password: 
 decrypt files for user admin (1 of 2): /admin/files/welcome.txt 
 [-->-------------------------]
Prepare "Default encryption module"

Configuring encryption module for decryption with user based keys
Please enter the user: user1 login password: 
 decrypt files for user user1 (2 of 2): /user1/files/welcome.txt 
 [--->------------------------]

 starting to decrypt files... finished 
 [============================]


all files could be decrypted successfully!
➜  owncloud3 git:(recovery_passwd-and-oc_passwd-from-env) ✗
```
2. Make some modification to `welcome.txt` files of both user and enable recovery password for both users, and observe the console log  
```
➜  owncloud3 git:(recovery_passwd-and-oc_passwd-from-env) ✗ ./occ encryption:decrypt-all -c yes -m recovery
Cannot load Xdebug - it was already loaded
Disable server side encryption... done.


You are about to start to decrypt all files stored in your ownCloud.
It will depend on the encryption module and your setup if this is possible.
Depending on the number and size of your files this can take some time
Please make sure that no user access his files during this process!

prepare encryption modules...
 done.


 starting to decrypt files... 
 [->--------------------------]
Prepare "Default encryption module"

Configuring encryption module for decryption with user based keys
Attempting to use recovery key from environment: OC_RECOVERY_PASSWORD
 decrypt files for user admin (1 of 2): /admin/files/welcome.txt 
 [-->-------------------------]
Prepare "Default encryption module"

Configuring encryption module for decryption with user based keys
Attempting to use recovery key from environment: OC_RECOVERY_PASSWORD
 decrypt files for user user1 (2 of 2): /user1/files/welcome.txt 
 [--->------------------------]

 starting to decrypt files... finished 
 [============================]


all files could be decrypted successfully!
➜  owncloud3 git:(recovery_passwd-and-oc_passwd-from-env) ✗
```
Now the files are observed decrypted.
3. Make some modification to `welcome.txt` files of both user, and observe the console log:
```
➜  owncloud3 git:(recovery_passwd-and-oc_passwd-from-env) ✗ ./occ encryption:decrypt-all -c yes -m password user1
Cannot load Xdebug - it was already loaded
Disable server side encryption... done.


You are about to start to decrypt all files stored in user1's account.
It will depend on the encryption module and your setup if this is possible.
Depending on the number and size of your files this can take some time
Please make sure that no user access his files during this process!

prepare encryption modules...
 done.


 starting to decrypt files... 
 [->--------------------------]
Prepare "Default encryption module"

Configuring encryption module for decryption with user based keys
Attempting to use users password from environment: OC_PASSWORD
 decrypt files for user user1 (1 of 1): /user1/files/welcome.txt 
 [-->-------------------------]

 starting to decrypt files... finished 
 [============================]


all files could be decrypted successfully!
Server side encryption remains enabled
➜  owncloud3 git:(recovery_passwd-and-oc_passwd-from-env) ✗ 
```
4. Make some modification to `welcome.txt` files of both user, and observe the console log:
```
➜  owncloud3 git:(recovery_passwd-and-oc_passwd-from-env) ✗ ./occ encryption:decrypt-all -c yes -m password      
Cannot load Xdebug - it was already loaded
Disable server side encryption... done.


You are about to start to decrypt all files stored in your ownCloud.
It will depend on the encryption module and your setup if this is possible.
Depending on the number and size of your files this can take some time
Please make sure that no user access his files during this process!

 aborted.

In DecryptAll.php line 172:
                                                            
  Password method cannot be used for decrypting all users.  
                                                            

encryption:decrypt-all [-m|--method [METHOD]] [-c|--continue [CONTINUE]] [--] [<user>]

➜  owncloud3 git:(recovery_passwd-and-oc_passwd-from-env) ✗
```
5. Make some modification to `welcome.txt` files of both user and disable recovery password for `user1`, and observe the console log:
```
➜  owncloud3 git:(recovery_passwd-and-oc_passwd-from-env) ✗ ./occ encryption:decrypt-all -c yes -m recovery
Cannot load Xdebug - it was already loaded
Disable server side encryption... done.


You are about to start to decrypt all files stored in your ownCloud.
It will depend on the encryption module and your setup if this is possible.
Depending on the number and size of your files this can take some time
Please make sure that no user access his files during this process!

prepare encryption modules...
 done.


 starting to decrypt files... 
 [->--------------------------]
Prepare "Default encryption module"

Configuring encryption module for decryption with user based keys
Attempting to use recovery key from environment: OC_RECOVERY_PASSWORD
 decrypt files for user admin (1 of 2): /admin/files/welcome.txt 
 [-->-------------------------]
Prepare "Default encryption module"

Configuring encryption module for decryption with user based keys
Password recovery is not enabled for user1
Module "Default encryption module" does not support the functionality to decrypt all files again or the initialization of the module failed!


 starting to decrypt files... finished 
 [============================]


all files could be decrypted successfully!
➜  owncloud3 git:(recovery_passwd-and-oc_passwd-from-env) ✗
```
For testing recovery `OC_RECOVERY_PASSWORD=xxx` and for password `OC_PASSWORD=xxx` is exported in the shell.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Technical debt
- [ ] Tests

## Checklist:

- [ ] Code changes
- [x] Unit tests added
- [ ] Acceptance tests added
- [ ] Documentation ticket raised: <link> 
- [x] REQUIRES: https://github.com/owncloud/encryption/pull/40
- [x] This PR was taken from https://github.com/owncloud/core/pull/28569, made minor modifications and added unit tests. Basically trying to take an approach of creating PR in the master branch first and then backport.

## Open tasks:

- [ ] Backport (if applicable set "backport-request" label and remove when the backport was done)

### CODE REVIEW

File: core/Command/Encryption/DecryptAll.php [hunk 1]
```diff
@@ -117,9 +118,27 @@ protected function configure() {
 			'User for whom you want to decrypt all files (optional).',
 			''
 		);
+		$this->addOption(
+			'method',
+			'm',
+			InputOption::VALUE_OPTIONAL,
+			'Use recovery or password. If recovery method is opted the recovery password would be used to decrypt. If password is opted user password would be opted.'
```
Comment (phil-davis) [inline core/Command/Encryption/DecryptAll.php]: Suggested text here:
```
Use recovery or password. If recovery method is chosen then the recovery password will be used to decrypt files. If password method is chosen then individual user passwords will be used to decrypt files.
```

**Comment by PVince81** [comment]
also I don't see in this PR diff where the environment variables are read

**Comment by sharidas** [comment]
> in what does this related to the other tasks ? missing link

Yah forgot to link the pr. Sorry for the mess. The pr in encryption app https://github.com/owncloud/encryption/pull/40 is where the reading from env happens.

**Comment by PVince81** [comment]
@sharidas your PR headers do not state that both PRs belong together, please adjust.

I usually use the format:
```
- [ ] REQUIRES: #link-to-other-ticket
```

File: core/Command/Encryption/DecryptAll.php [hunk 2]
```diff
@@ -144,9 +163,14 @@ protected function execute(InputInterface $input, OutputInterface $output) {
...
 			$output->writeln('');
 			$question = new ConfirmationQuestion('Do you really want to continue? (y/n) ', false);
-			if ($this->questionHelper->ask($input, $output, $question)) {
+			if (($confirmed === 'yes') || $this->questionHelper->ask($input, $output, $question)) {
 				$this->forceSingleUserAndTrashbin();
 				$user = $input->getArgument('user');
+				//Try to stop decrypting process if password method is used for decrypting all users.
+				if (($user === '') && ($input->getOption('method') === 'password')) {
+					$output->writeln(' aborted.');
+					throw new \Exception('Password method cannot be used for decrypting all users.');
```
Comment (PVince81) [inline core/Command/Encryption/DecryptAll.php]: shouldn't this be done in the default encryption module instead ?

as I understand core doesn't have a concept of "method" argument

Comment (sharidas) [reply core/Command/Encryption/DecryptAll.php]: Moved this to default encryption module.

File: tests/lib/Helper/EnvironmentHelperTest.php [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+class EnvironmentHelperTest extends TestCase {
+	private $environmentHelper;
+
+	public function setUp() {
+		$this->environmentHelper = new EnvironmentHelper();
+		parent::setUp();
+	}
+
+	public function testGetWebRoot() {
+		$this->assertEquals(\OC::$WEBROOT, $this->environmentHelper->getWebRoot());
```
Comment (PVince81) [inline tests/lib/Helper/EnvironmentHelperTest.php]: is this even defined when running unit tests ?

you likely need to set the value before the call, then reset it to its former value after assertion

**Review by PVince81 — APPROVED** [review]
:+1:

**Comment by PVince81** [comment]
@sharidas please backport

**Comment by mmattel** [comment]
This is documentation relevant ! Please open a doc ticket !
``encryption:decrypt-all user --method= --continue=``

**Comment by sharidas** [comment]
Backport PR -> https://github.com/owncloud/core/pull/32252

**Comment by PVince81** [comment]
potential regression: https://github.com/owncloud/encryption/issues/51

**Pull request merged**